src/tracks/playabletrack/wavetrack/ui/WaveTrackShifter.cpp

```diff
@@ -52,8 +66,12 @@ class WaveTrackShifter final : public TrackShifter {
          auto data =
             static_cast<WaveTrack::IntervalData*>( myInterval.Extra() );
          auto clip = data->GetClip().get();
-         return !(clip->IsClipStartAfterClip(interval.Start()) ||
-            clip->BeforeClip(interval.End()));
+         const auto c0 = mpTrack->TimeToLongSamples(clip->GetStartTime());
+         //see comment for HitTest routine
+         const auto c1 = mpTrack->TimeToLongSamples(clip->GetEndTime() - 1.0 / mpTrack->GetRate());
+         return 
+             mpTrack->TimeToLongSamples(interval.Start()) <= c1 && 
+             mpTrack->TimeToLongSamples(interval.End()) >= c0;
```

Comment:
Now I understand!  WaveClip::BeforeClip() is really the faulty function.  It contains <= where it should have < .

Here, you just avoid the use of BeforeClip().

There are uses of BeforeClip() elsewhere, though, which should be reexamined, and may reveal more bugs like this one.

Reply:
>Even if it is not different, I agree it's good to write similar tests in the two places in this file.

Yes, it is the root of the issue - clip borders were determined differently in clicked track and other channels. Another issue is that visual borders of the clip do not correspond to a actual clip borders:
![image](https://user-images.githubusercontent.com/20165230/134674196-6aab2eb9-2cb8-4e89-943d-d27d068595be.png)
Here selection shows the actual length of the clip, but frame is drawn with offset and a bit shorter, to prevent frame overlap with samples and overlap between adjacent clips if there is no gap between them. I think that the drawing is done wrong https://github.com/audacity/audacity/issues/1689#issuecomment-926428948 The solution might be to draw frame to exactly match the beginning and the end of a clip, but move "dot" sample by half of a sample.

Reply:
How about exploring the uses of BeforeClip() for other off-by-one bugs that we have not yet identified?